### PR TITLE
LPD-17894, LPD-19243 Translations for Web Content

### PR DESCRIPTION
**Pull request by NemethNorbert**
[LPD-19243](https://liferay.atlassian.net/browse/LPD-19243)
[LPD-17894](https://liferay.atlassian.net/browse/LPD-17894)

Previous PR: https://github.com/liferay-content-management/liferay-portal/pull/5173

I resent this PR with LPD-19243 as it depends on LPD-17894.

Changes I applied to LPD-17894:
https://github.com/liferay-content-management/liferay-portal/commit/b01a1fd4512b306c06579684d7d5d0dd682b95b5 : The TranslationManager was not able to track rich-text localisation due to a bug, this is the fix for it.

https://github.com/liferay-content-management/liferay-portal/commit/dc4ff026be5a7d364b575792f6400b80da66bd70 : We should delete the translation instead of setting the value to the default locale's value, so data-engine knows it's not localised

https://github.com/liferay-content-management/liferay-portal/commit/872a3477be4bde7c934e41e1d9b748b1569ef66d : Once the above is done, we need the event to update the ddmFields

https://github.com/liferay-content-management/liferay-portal/commit/e841939af34f8213b79ff19235e69bd4508a9331 : rich-text requires additional care to update the editor's value, it's needed as the component tracks the value internally and we cannot force an update on the editor itself from outside the component

https://github.com/liferay-content-management/liferay-portal/commit/07b58226c67ac541656785d0d1b7ed0a8488c46a : We did not handle the defaultLocaleChanged event, this is a fix for that -> I don't like the typescript handling of that event, we might need to create a type for it to handle it properly

### CODE REVIEW

**Comment by NemethNorbert** [comment]
ci:test:sf

**Review by ambrinchaudhary — CHANGES_REQUESTED** [review]
I'm still a bit lost with all DDM fields and events involved, so I didn't pay much attention to that. 
I have some other minor changes request, please fell free to correct me if I misunderstood something. 
I see there is a lot of work here and I guess it was a complex work, well done 💪

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationOptions.tsx [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+					<ClayModal.Header>
+						{sub(
+							Liferay.Language.get('mark-x-as-translated'),
+							selectedLanguageId
+						)}
+					</ClayModal.Header>
+
+					<ClayModal.Body>
+						<p
+							dangerouslySetInnerHTML={{
```
[the remark this answers is not in the document]
Comment (DiegoHu97) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationOptions.tsx:221]: Hi @ambrinchaudhary , about this, I think the SF fails if we add html into it

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationOptions.tsx [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+	const [dropdownActive, setDropdownActive] = useState(false);
+
+	const markAsTranslatedHandler = () => {
+		Liferay.fire('inputLocalized:markAsTranslated', {selectedLanguageId});
+	};
+
+	const resetButtonHandler = () => {
+		const fields = getAllLocalizableFields(initialFields);
+
+		const defaultFields = [
```
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationOptions.tsx]: I decided to remove it, as we only need to reset the initialFields manually

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+				<TranslationManager
+					defaultLanguageId={defaultLanguageId}
+					fields={fields}
+					getLocalizableFields={getLocalizableFields}
+					locales={locales}
+					namespace={namespace}
+					selectedLanguageId={selectedLanguageId}
+					setFields={setFields}
+					setSelectedLanguageId={setSelectedLanguageId}
+					setTranslations={setTranslations}
```
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx]: Thank you @ambrinchaudhary ,

I decided to create a hook from the TranslationWrapper component to hopefully make the component more readable as well as decouple them from the layout dependancy its created

File: modules/apps/journal/journal-web/tsconfig.json [hunk 4]
```diff
@@ -1,6 +1,12 @@
 {
-	"@generated": "f16cae6bfb35ec5c5081794ee82f5a0b65a8a373",
+	"@generated": "eafdaa32eaaaa1ea2d06f922bcd722233127d870",
 	"@overrides": {
+		"compilerOptions": {
```
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/tsconfig.json:4]: @DiegoHu97 ,
 Can you help with an answer here?

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+			`[data-ddm-localizable-field-id]`
+		)
+	)
+		.map(
+			(field) =>
+				`${field.dataset.fieldName}${field.dataset.ddmLocalizableFieldId}`
+		)
+		.reduce((acc, name) => ({...acc, [name]: {}}), {});
+
+	return {...initialFields, ...ddmFields};
```
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx]: Yes it does, I use this in the hook I created from the wrapper

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+			? (Object.keys(fields[fieldName]) as Liferay.Language.Locale[])
+			: [];
+
+		translations.push({
+			fieldName,
+			languages,
+		});
+	}
+
+	return translations;
```
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx]: Yes it does, I use this in the hook I created from the wrapper

File: modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+export default function TranslationsWrapper({
+	defaultLanguageId: currentDefaultLanguageId,
+	fields: initialFields,
+	locales,
+	namespace,
+	selectedLanguageId: initialSelectedLanguageId,
+}: TranslationsWrapper) {
+	const [defaultLanguageId, setDeafultLanguageId] = useState(
```
[the remark this answers is not in the document]
Comment (NemethNorbert) [reply modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/translation_manager/TranslationsWrapper.tsx]: Thanks, I fixed the type, now it's in the hook

File: modules/apps/journal/journal-web/tsconfig.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ambrinchaudhary) [reply modules/apps/journal/journal-web/tsconfig.json:4]: hi @DiegoHu97, any update about this? I would try to remove them to check if everything works, I don't think those libs are necessary...

**Comment by ambrinchaudhary** [comment]
Just pending confirmation about https://github.com/liferay-content-management/liferay-portal/pull/5193#discussion_r1519574502, the rest LGTM.

**Comment by DiegoHu97** [comment]
Hey @NemethNorbert , there are some conflicts, could you takel a look to it? thanks! :smile:

**Comment by NemethNorbert** [comment]
> Hey @NemethNorbert , there are some conflicts, could you takel a look to it? thanks! 😄

I rebased and force pushed the branch, the conflict was with one of the language file

**Comment by NemethNorbert** [comment]
Another push as I realised I used the wron LPD number on the last commit

**Comment by NemethNorbert** [comment]
Hey @ambrinchaudhary, @DiegoHu97 
I just noticed a rather big issue with the hook. The translations are beeing recalculated every time a change is registered in any on the input_localized field. I have to fix it before we move forward it. I started working on it

**Comment by NemethNorbert** [comment]
It's fixed, we fired the updateTranslationStatus event everytime an inputLocalized field was updated. It is not needed, so I removed the event

File: modules/apps/journal/journal-web/tsconfig.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ambrinchaudhary) [reply modules/apps/journal/journal-web/tsconfig.json:4]: I spoke to @DiegoHu97, and he commented that those libs are neccessary for the `.flatMap` function used in `ResourceSelector.tsx`

**Review by ambrinchaudhary — APPROVED** [review]

**Comment by ambrinchaudhary** [comment]
ci:forward

**Comment by ambrinchaudhary** [comment]
ci:forward

**Comment by DiegoHu97** [comment]
ci:report:2550021467

**Comment by ambrinchaudhary** [comment]
For me errors do not look relevant to this pull... anyway I will try tu run ci:test:relevant again to confirm if errors are not repeated.

**Comment by ambrinchaudhary** [comment]
ci:test:relevant

**Comment by ambrinchaudhary** [comment]
ci:forward:force

**Comment by DiegoHu97** [comment]
ci:forward:force

**Comment by ambrinchaudhary** [comment]
Hey Diego, probably this error is related to the pull: 

> Execution failed for task ':apps:journal:journal-web:packageRunTest'

**Comment by DiegoHu97** [comment]
Updates TranslationManager.test.tsx and resent here: #5254   
CC: @ambrinchaudhary @NemethNorbert

**Pull request closed without merging**